File: db-esdk-performance-testing/benchmarks/go/benchmark/results.go
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package benchmark

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"time"
)

// BenchmarkResult represents the results of a single benchmark test
type BenchmarkResult struct {
	TestName          string  `json:"test_name"`
	Language          string  `json:"language"`
	DataSize          int     `json:"data_size"`
	Concurrency       int     `json:"concurrency"`
	PutLatencyMs      float64 `json:"put_latency_ms"`
	GetLatencyMs      float64 `json:"get_latency_ms"`
	EndToEndLatencyMs float64 `json:"end_to_end_latency_ms"`
	OpsPerSecond      float64 `json:"ops_per_second"`
	BytesPerSecond    float64 `json:"bytes_per_second"`
	PeakMemoryMB      float64 `json:"peak_memory_mb"`
	MemoryEfficiency  float64 `json:"memory_efficiency_ratio"`
	P50Latency        float64 `json:"p50_latency"`
	P95Latency        float64 `json:"p95_latency"`
	P99Latency        float64 `json:"p99_latency"`
	Timestamp         string  `json:"timestamp"`
	GoVersion         string  `json:"go_version"`
	CPUCount          int     `json:"cpu_count"`
	TotalMemoryGB     float64 `json:"total_memory_gb"`
}

// MemorySample represents a single memory measurement
type MemorySample struct {
	Timestamp        time.Time
	HeapMB           float64
	MetricsAllocsMB  float64
	MemStatsAllocsMB float64
}

// === Utility Functions ===

// Average calculates the average of a slice of float64 values
func Average(values []float64) float64 {
	if len(values) == 0 {
		return 0
	}
	sum := 0.0
	for _, v := range values {
		sum += v
	}
	return sum / float64(len(values))
}

// Percentile calculates the percentile of sorted values
func Percentile(sortedValues []float64, p float64) float64 {
	if len(sortedValues) == 0 {
		return 0
	}
	if p <= 0 {
		return sortedValues[0]
	}
	if p >= 100 {
		return sortedValues[len(sortedValues)-1]
	}

	index := (p / 100.0) * float64(len(sortedValues)-1)
	lower := int(math.Floor(index))
	upper := int(math.Ceil(index))

	if lower == upper {
		return sortedValues[lower]
	}

	weight := index - float64(lower)
	return sortedValues[lower]*(1-weight) + sortedValues[upper]*weight
}

// === Results Saving ===

// SaveResults saves benchmark results to JSON file
func (b *DBESDKBenchmark) SaveResults(outputPath string) error {
	if err := os.MkdirAll(filepath.Dir(outputPath), 0755); err != nil {
		return fmt.Errorf("failed to create output directory: %w", err)
	}

	resultsData := map[string]interface{}{
		"metadata": map[string]interface{}{
			"language":        "go",
			"timestamp":       time.Now().Format("2006-01-02 15:04:05"),
			"go_version":      runtime.Version(),
			"cpu_count":       b.CPUCount,
			"total_memory_gb": b.TotalMemoryGB,
			"total_tests":     len(b.Results),
		},
		"results": b.Results,
	}

	file, err := os.Create(outputPath)
	if err != nil {
		return fmt.Errorf("failed to create output file: %w", err)
	}
	defer file.Close()

	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(resultsData); err != nil {
		return fmt.Errorf("failed to encode results to JSON: %w", err)
	}

	return nil
}

File: db-esdk-performance-testing/benchmarks/rust/src/results.rs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::alloc;
use anyhow::Result;
use chrono::Utc;
use serde::Serialize;
use std::fs::{self, File};
use std::io::Write;
use std::path::Path;

#[derive(Debug, Serialize, Clone)]
pub struct BenchmarkResult {
    pub test_name: String,
    pub language: String,
    pub data_size: usize,
    pub concurrency: usize,
    pub encrypt_latency_ms: f64,
    pub decrypt_latency_ms: f64,
    pub end_to_end_latency_ms: f64,
    pub ops_per_second: f64,
    pub bytes_per_second: f64,
    pub peak_memory_mb: f64,
    pub memory_efficiency_ratio: f64,
    pub p50_latency: f64,
    pub p95_latency: f64,
    pub p99_latency: f64,
    pub timestamp: String,
    pub rust_version: String,
    pub cpu_count: usize,
    pub total_memory_gb: f64,
    pub alloc: alloc::ResourceResults,
}

#[derive(Debug, Serialize)]
struct BenchmarkResults {
    metadata: BenchmarkMetadata,
    results: Vec<BenchmarkResult>,
}

#[derive(Debug, Serialize)]
struct BenchmarkMetadata {
    language: String,
    timestamp: String,
    rust_version: String,
    cpu_count: usize,
    total_memory_gb: f64,
    total_tests: usize,
}

// === Utility Functions ===

pub fn average(values: &[f64]) -> f64 {
    if values.is_empty() {
        return 0.0;
    }
    values.iter().sum::<f64>() / values.len() as f64
}

pub fn percentile(sorted_values: &[f64], p: f64) -> f64 {
    if sorted_values.is_empty() {
        return 0.0;
    }
    if p <= 0.0 {
        return sorted_values[0];
    }
    if p >= 100.0 {
        return sorted_values[sorted_values.len() - 1];
    }

    let index = (p / 100.0) * (sorted_values.len() - 1) as f64;
    let lower = index.floor() as usize;
    let upper = index.ceil() as usize;

    if lower == upper {
        return sorted_values[lower];
    }

    let weight = index - lower as f64;
    sorted_values[lower] * (1.0 - weight) + sorted_values[upper] * weight
}

// === Results Saving ===

pub fn save_results(
    results: &[BenchmarkResult],
    output_path: &str,
    cpu_count: usize,
    total_memory_gb: f64,
) -> Result<()> {
    // Create output directory if it doesn't exist
    if let Some(parent) = Path::new(output_path).parent() {
        fs::create_dir_all(parent)?;
    }

    let results_data = BenchmarkResults {
        metadata: BenchmarkMetadata {
            language: "rust".to_string(),
            timestamp: Utc::now().format("%Y-%m-%d %H:%M:%S").to_string(),
            rust_version: std::env::var("RUSTC_VERSION").unwrap_or_else(|_| "unknown".to_string()),
            cpu_count,
            total_memory_gb,
            total_tests: results.len(),
        },
        results: results.to_vec(),
    };

    let json_data = serde_json::to_string_pretty(&results_data)?;
    let mut file = File::create(output_path)?;
    file.write_all(json_data.as_bytes())?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_average() {
        let values = vec![1.0, 2.0, 3.0, 4.0, 5.0];
        assert_eq!(average(&values), 3.0);
        
        let empty: Vec<f64> = vec![];
        assert_eq!(average(&empty), 0.0);
    }
    
    #[test]
    fn test_percentile() {
        let mut values = vec![1.0, 2.0, 3.0, 4.0, 5.0];
        values.sort_by(|a, b| a.partial_cmp(b).unwrap());
        
        assert_eq!(percentile(&values, 50.0), 3.0);
        assert_eq!(percentile(&values, 0.0), 1.0);
        assert_eq!(percentile(&values, 100.0), 5.0);
        
        let empty: Vec<f64> = vec![];
        assert_eq!(percentile(&empty, 50.0), 0.0);
    }
    
    #[test]
    fn test_format_data_size() {
        assert_eq!(format_data_size(512), "512 B");
        assert_eq!(format_data_size(1024), "1.0 KB");
        assert_eq!(format_data_size(1024 * 1024), "1.0 MB");
        assert_eq!(format_data_size(1024 * 1024 * 1024), "1.0 GB");
    }
}
